app/models/project_services/composer_service.rb

```diff
@@ -0,0 +1,459 @@
+# == Schema Information
+#
+# Table name: services
+#
+#  id         :integer          not null, primary key
+#  type       :string(255)
+#  title      :string(255)
+#  project_id :integer          not null
+#  created_at :datetime
+#  updated_at :datetime
+#  active     :boolean          default(FALSE), not null
+#  properties :text
+#
+
+require 'digest/crc32'
+
+class ComposerService < Service
+
+  prop_accessor :package_mode, :package_type, :export_branches, :branch_filters, :export_tags, :tag_filters, :custom_json
+
+  validates :package_mode, :package_type, presence: true, if: :activated?
+  validates :custom_json,
+    presence: true,
+    if: ->(service) { service.activated? && service.package_mode == 'advanced' }
+
+  validates_each :custom_json,
+    if: :allow_custom_json_validation? do |record, attr, value|
+      begin
+        if value.empty?
+          record.errors.add(attr, 'must be specified')
+        elsif (cjson = ActiveSupport::JSON.decode(value))
+          if cjson.empty?
+            record.errors.add(attr, 'must not be empty')
+          elsif not cjson["name"]
+            record.errors.add(attr, 'must have a name key specified')
+          elsif cjson["name"].empty?
+            record.errors.add(attr, 'name key must not be empty')
+          elsif cjson["name"] != (/([A-Za-z0-9&_-]+\/[A-Za-z0-9&_-]+)/.match(cjson["name"]) || [])[0]
```

Comment:
Line is too long. [101/80]<br>Prefer single-quoted strings when you don't need string interpolation or special symbols.